src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.Windows.cs

```diff
@@ -113,53 +82,23 @@ private void InitFromHandle(SafeFileHandle handle, FileAccess access, bool useAs
 
         private void InitFromHandleImpl(SafeFileHandle handle, bool useAsyncIO)
         {
-            FileStreamHelpers.GetFileTypeSpecificInformation(handle, out _canSeek, out _isPipe);
-
-            // This is necessary for async IO using IO Completion ports via our
-            // managed Threadpool API's.  This calls the OS's
-            // BindIoCompletionCallback method, and passes in a stub for the
-            // LPOVERLAPPED_COMPLETION_ROUTINE.  This stub looks at the Overlapped
-            // struct for this request and gets a delegate to a managed callback
-            // from there, which it then calls on a threadpool thread.  (We allocate
-            // our native OVERLAPPED structs 2 pointers too large and store EE
-            // state & a handle to a delegate there.)
-            //
-            // If, however, we've already bound this file handle to our completion port,
-            // don't try to bind it again because it will fail.  A handle can only be
-            // bound to a single completion port at a time.
-            if (useAsyncIO && !(handle.IsAsync ?? false))
-            {
-                try
-                {
-                    handle.ThreadPoolBinding = ThreadPoolBoundHandle.BindHandle(handle);
-                }
-                catch (Exception ex)
-                {
-                    // If you passed in a synchronous handle and told us to use
-                    // it asynchronously, throw here.
-                    throw new ArgumentException(SR.Arg_HandleNotAsync, nameof(handle), ex);
-                }
-            }
-            else if (!useAsyncIO)
-            {
-                FileStreamHelpers.VerifyHandleIsSync(handle);
-            }
+            handle.InitThreadPoolBindingIfNeeded();
```

Comment:
InitThreadPoolBindingIfNeeded disposes of the handle in some cases.  That wasn't being done before, by design, since we didn't create the handle.